[RFC] Update cached_status_data column 1/2

clairelin135 opened the pull request:
Following the [cached asset partition status](https://www.notion.so/dagster/Cached-incremental-asset-partition-summaries-9afd9aa0b91940d49c3d5af5a832131a) proposal, this PR creates an `update_asset_cached_status_data` method that writes to the new cached_status_data column of the AssetKeyTable. This column now contains:

- latest evaluated storage ID of the asset
- a serialized unique identifier to represent the partitions definition. If the partitions definition is changed, the mismatch between the serialized identifier and the new identifier triggers a recalculation and update to the cached status column.
- serialized representation of the materialized partition subset. This contains partition ranges for time window partitions definitions, and a set of materialized partition keys otherwise.

The `update_asset_cached_status_data` will check if the column exists, and if it does, it will:
- read the currently cached data per asset and
- if the partition definition unique identifier differs, recalculates the partition materialization status from event log storage
- otherwise, the unique identifier remains the same and the method reads the unevaluated events
- updates the `cached_status_data` column with the new partition materialization data

PR for part 2 updates Dagit to read from the cached data: https://github.com/dagster-io/dagster/pull/10822

sryza commented on python_modules/dagster/dagster/_core/storage/partition_status_cache.py:
```diff
@@ -0,0 +1,273 @@
+        return partitions_def.deserialize_subset(self.materialized_partition_subsets)
+
+
+def get_materialized_multipartitions_from_tags(
```
Nitpick: I think could leave `from_tags` out, because that's an implementation detail. This name led me to think that the tags would be one of the arguments.

sryza commented on python_modules/dagster/dagster/_core/storage/partition_status_cache.py:
```diff
@@ -0,0 +1,273 @@
+        }
+
+        if all(
+            [
```
Can take out these square brackets - `all` can operate directly on iterables without wrapping them in lists.

sryza commented on python_modules/dagster/dagster/_core/storage/partition_status_cache.py:
```diff
@@ -0,0 +1,273 @@
+    return materialized_keys
+
+
+def _rebuild_status_cache(
```
Nitpick: this won't always be a _re_-build, right? I.e. we'll call this the first time? If so, maybe a name like "build from scratch"?

sryza commented on python_modules/dagster/dagster/_core/storage/partition_status_cache.py:
```diff
@@ -0,0 +1,273 @@
+        [
+            ("latest_storage_id", int),
+            ("partitions_def_id", Optional[str]),
+            ("materialized_partition_subsets", Optional[str]),
```
My initial reaction was to think it's a little weird to represent this property as serialized instead of deserializing it and holding the `PartitionsSubset`. I suppose the advantage of having it serialized though is that we can avoid the deserialization cost in cases where only we need other entries from the cache?  If we do want to have it serialized, we should include that in the name to make it clear.

Also, subsets should be singular, not plural, right? I.e. it's just one subset, which includes multiple partitions?

clairelin135 replied:
The problem is that we need the partitions definition that was used to serialize the partitions subset in order to know how to deserialize it. For example, given a serialized partitions subset, we don't know if it is a static partitions definition or a multi-partitions definition.

I figured that since we only need the cached value if the partitions definition remains the same, we can just deserialize it in those cases. I'll rename this to `serialized_materialized_partition_subset`.

sryza commented on python_modules/dagster/dagster/_core/storage/partition_status_cache.py:
```diff
@@ -0,0 +1,273 @@
+        partitions_def_id (Optional(str)): The serializable unique identifier for the partitions
+            definition. When this value differs from the new partitions definition, this cache
+            value needs to be recalculated. None if the asset is unpartitioned.
+        materialized_partition_subsets (Optional(str)): The serializable representation of the
```
```suggestion
        materialized_partition_subsets (Optional(str)): The serialized representation of the
```

clairelin135 commented on python_modules/dagster/dagster/_core/storage/partition_status_cache.py:
```diff
@@ -0,0 +1,273 @@
+    the materialized partition subset for the asset key and updates the cache value.
+    """
+    if not partitions_def:
+        return AssetStatusCacheValue(latest_storage_id=latest_storage_id)
```
I think for un-partitioned assets we still want to store an asset status cache value. This will improve performance for things like fetching freshness data via graphql in batches since we currently don't have access to the latest storage id via the asset key table

clairelin135 commented:
@sryza I've updated this PR per your feedback. I've also added a set of changes so this feature should be fully functional now:
- Wiping an asset will now also wipe its cached data
- Disallowing dynamic partitions since we don't display them in Dagit
- Improved test coverage covering each possible partition type

sryza commented on python_modules/dagster/dagster/_core/definitions/multi_dimensional_partitions.py:
```diff
@@ -221,6 +231,53 @@ def __str__(self) -> str:
+            f"but got {len(partition_key_strs)}",
+        )
+
+        all_partition_keys = self.get_partition_keys()
```
I think it will be pretty expensive to get all the partition keys every time this method is called.

sryza reviewed (approved):
This looks great.

A couple small things before merging:
- Maybe it's already there and I didn't read closely enough, but can we make sure there's a test that asserts that the cache gets discarded if the PartitionsDefinition changes?
- I left a comment about perf

prha commented on python_modules/dagster/dagster/_core/storage/event_log/sql_event_log.py:
```diff
@@ -1126,14 +1152,8 @@ def _fetch_raw_asset_rows(self, asset_keys=None, prefix=None, limit=None, cursor
-                    AssetKeyTable.c.tags,
-                ]
-            )
+        if self.has_asset_key_col("cached_status_data"):
```
nit: should this use `can_cache_asset_status_data`?

prha reviewed (approved):
Nice job on this
